StudentDriver/StudentDriver/DrivePage.xaml.cs

```diff
@@ -1,24 +1,145 @@
 ﻿using System;
 using System.Collections.Generic;
-
+using System.Threading.Tasks;
+using System.Diagnostics;
+using Plugin.Geolocator;
+using Plugin.Geolocator.Abstractions;
 using Xamarin.Forms;
+using StudentDriver.Services;
+using StudentDriver.Models;
 
 namespace StudentDriver
 {
 	public partial class DrivePage : ContentPage
 	{
-		private bool drivingButtonPressed = false;
-		public DrivePage ()
+		private bool isStudentDriving = false;
+		//speed and distance in miles (imperial)
+		private double currentAverageSpeed = 0.0;
+		private TimeSpan currentTime;
+		private List<DrivePoint> positions = new List<DrivePoint>();
+		private IGeolocator locator;
+
+		protected override void OnAppearing()
 		{
-			InitializeComponent ();
-			NavigationPage.SetHasNavigationBar (this, false);
-			drivingButton.Clicked += async (object sender, EventArgs e) => {
-				drivingButton.BackgroundColor = drivingButtonPressed ? AppColors.Third : AppColors.Fourth;
-				drivingButtonPressed = !drivingButtonPressed;
-				drivingButton.Text = drivingButtonPressed ? "Stop" : "Start";
+			base.OnAppearing();
+
+			try
+			{
+				locator = CrossGeolocator.Current;
+				locator.DesiredAccuracy = 5;
+				locator.PositionChanged += (object s, PositionEventArgs eventArg) =>
+				{
+					var currentPosition = eventArg.Position;
+					if (positions.Count < 100) // every 100 points, we will add to the database.
+					{
+						var drivePoint = new DrivePoint();
+						drivePoint.Latitude = currentPosition.Latitude;
+						drivePoint.Longitude = currentPosition.Longitude;
+						drivePoint.PointDateTime = currentPosition.Timestamp.DateTime;
+						drivePoint.Speed = (float)currentPosition.Speed;
+						positions.Add(drivePoint);
+						currentAverageSpeed = currentAverageSpeed == 0.0 ? ConvertSpeeds(eventArg.Position.Speed) : (0.95 * currentAverageSpeed) + (0.05 * ConvertSpeeds(eventArg.Position.Speed));
+					}
+					else
+					{
+						Task.Factory.StartNew(async () =>
+						{
+							var database = SQLiteDatabase.GetInstance();
+							await database.AddDrivePoints(positions);
+							positions.Clear();
+						});
+					}
+					avgSpeedLabel.Text = string.Format("{0} MPH", currentAverageSpeed.ToString("F1"));
```

Comment:
@zawadzkip @DylanWalseth Offhand, do any of you know a .js lib for processing GPS points to calculate speed (that accounts for factors such as elevations, etc. iirc we also discussed about this and our experience with the GPS lab in SE2800/Software Processes in Sophomore Year)? I remember past discussions that it would make sense for the backend/server to handle all the legwork for mathematical calculations so that the mobile app is fast/responsive for the end-user.